src/java/org/apache/cassandra/db/filter/ClusteringIndexNamesFilter.java

```diff
@@ -161,34 +162,41 @@ public String toString(TableMetadata metadata)
         return sb.append(')').toString();
     }
 
-    public String toCQLString(TableMetadata metadata)
+    @Override
+    public String toCQLString(TableMetadata metadata, RowFilter rowFilter, boolean redact)
     {
         if (metadata.clusteringColumns().isEmpty() || clusterings.isEmpty())
-            return "";
+            return rowFilter.toCQLString(redact);
 
-        StringBuilder sb = new StringBuilder();
+        boolean isSingleColumn = metadata.clusteringColumns().size() == 1;
+        boolean isSingleClustering = clusterings.size() == 1;
 
-        boolean multipleColumns = metadata.clusteringColumns().size() > 1;
-        boolean multipleClusterings = clusterings.size() > 1;
+        StringBuilder sb = new StringBuilder();
+        sb.append(isSingleColumn ? "" : '(')
+          .append(ColumnMetadata.toCQLString(metadata.clusteringColumns()))
+          .append(isSingleColumn ? "" : ')');
 
-        if (multipleColumns)
-            sb.append('(');
-        sb.append(ColumnMetadata.toCQLString(metadata.clusteringColumns()));
-        if (multipleColumns)
-            sb.append(')');
-        sb.append(multipleClusterings ? " IN (" : " = ");
+        sb.append(isSingleClustering ? " = " : " IN (");
         int i = 0;
         for (Clustering<?> clustering : clusterings)
         {
-            sb.append(i++ == 0 ? "" : ", ");
-            if (multipleColumns)
-                sb.append('(');
-            sb.append(clustering.toCQLString(metadata));
-            if (multipleColumns)
-                sb.append(')');
+            sb.append(i++ == 0 ? "" : ", ")
+              .append(isSingleColumn ? "" : '(')
+              .append(clustering.toCQLString(metadata, redact))
+              .append(isSingleColumn ? "" : ')');
+
+            for (int j = 0; j < clustering.size(); j++)
+                rowFilter = rowFilter.withoutFirstLevelExpression(metadata.clusteringColumns().get(j), Operator.EQ, clustering.bufferAt(j));
+        }
+        sb.append(isSingleClustering ? "" : ")");
+        rowFilter = rowFilter.withoutFirstLevelExpression(metadata.clusteringColumns().get(0), Operator.IN);
+
+        if (!rowFilter.isEmpty())
+        {
+            String filter = rowFilter.toCQLString(redact);
+            sb.append(filter.startsWith("ORDER BY") ? " " : " AND ");
+            sb.append(filter);
```

Comment:
This logic is repeated in different places as rowFilter might contain `order by`. Does it make sense to have an utility method?